pkg/controller/machineset/machineset_controller_suite_test.go

```diff
@@ -17,59 +17,101 @@ limitations under the License.
 )
 
-var cfg *rest.Config
+func init() {
+	klog.InitFlags(nil)
+	logf.SetLogger(klogr.New())
 
-func TestMain(m *testing.M) {
-	t := &envtest.Environment{
+	// Register required object kinds with global scheme.
+	_ = machinev1.AddToScheme(scheme.Scheme)
+}
+
+const (
+	timeout = time.Second * 10
+)
+
+var (
+	cfg        *rest.Config
+	k8sClient  client.Client
+	testEnv    *envtest.Environment
+	mgr        manager.Manager
+	doneMgr    = make(chan struct{})
+	ctx        = context.Background()
+	reconciler reconcile.Reconciler
+)
+
+func TestMachinesetController(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"MachineSet Controller Suite",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = BeforeSuite(func(done Done) {
+	By("bootstrapping test environment")
+	testEnv = &envtest.Environment{
 		CRDDirectoryPaths: []string{filepath.Join("..", "..", "..", "install")},
 	}
-	v1beta1.AddToScheme(scheme.Scheme)
 
 	var err error
-	if cfg, err = t.Start(); err != nil {
-		log.Fatal(err)
-	}
+	cfg, err = testEnv.Start()
+	Expect(err).ToNot(HaveOccurred())
+	Expect(cfg).ToNot(BeNil())
 
-	code := m.Run()
-	t.Stop()
-	os.Exit(code)
-}
+	By("setting up a new manager")
+	mgr, err := manager.New(cfg, manager.Options{MetricsBindAddress: "0"})
+	Expect(err).NotTo(HaveOccurred())
 
-// SetupTestReconcile returns a reconcile.Reconcile implementation that delegates to inner and
-// writes the request to requests after Reconcile is finished.
-func SetupTestReconcile(inner reconcile.Reconciler) (reconcile.Reconciler, chan reconcile.Request) {
-	requests := make(chan reconcile.Request)
-	fn := reconcile.Func(func(req reconcile.Request) (reconcile.Result, error) {
-		result, err := inner.Reconcile(req)
-		requests <- req
-		return result, err
-	})
-	return fn, requests
-}
+	k8sClient = mgr.GetClient()
 
-// StartTestManager adds recFn
-func StartTestManager(mgr manager.Manager, t *testing.T) (chan struct{}, chan error) {
-	t.Helper()
+	By("setting up a new reconciler")
+	r := newReconciler(mgr)
+	reconciler = SetupTestReconcile(r)
 
-	stop := make(chan struct{})
-	errs := make(chan error, 1)
+	err = add(mgr, reconciler, r.MachineToMachineSets)
+	Expect(err).NotTo(HaveOccurred())
 
+	By("starting the manager")
 	go func() {
-		errs <- mgr.Start(stop)
+		Expect(mgr.Start(doneMgr)).To(Succeed())
 	}()
 
-	return stop, errs
+	close(done)
+}, 60)
+
+var _ = AfterSuite(func() {
+	By("closing the manager")
+	close(doneMgr)
+	By("tearing down the test environment")
+	Expect(testEnv.Stop()).To(Succeed())
+})
+
+// SetupTestReconcile returns a reconcile.Reconcile implementation that delegates to inner and
+// writes the request to requests after Reconcile is finished.
+func SetupTestReconcile(inner reconcile.Reconciler) reconcile.Reconciler {
+	fn := reconcile.Func(func(req reconcile.Request) (reconcile.Result, error) {
+		result, err := inner.Reconcile(req)
+		return result, err
+	})
+	return fn
```

Comment:
I don't think this actually does anything for us? I've seen this pattern before where it writes to a channel, but I always found relying on that to be annoying and not help us to write good tests, my suggestion would be to remove this

Reply:
👍 I thought about removing it too